homeassistant/components/nut/services.yaml

```diff
@@ -0,0 +1,18 @@
+run_command:
+  # Service name as shown in UI
+  name: Run command
+  # Description of the service
+  description: Invoke a given instant command in the UPS unit
+  fields:
+    # Key of the field
+    name:
+      # Field name as shown in UI
+      name: Name
+      # Description of the field
+      description: Command name
+      # Whether or not field is required (default = false)
+      required: true
+      # Example value that can be passed for this field
+      example: "beeper.disable"
+      # The default field value
+      default: ""
```

Comment:
Is it possible to map the commands to individual services? Like rather then having a single `run_command` service that requires users go look in the NUT documentation to see what values are allowed for `name`, could we have `nut.beeper_disable`, `nut.beeper_enable`, `nut.beeper_on`, etc. services?

Also what about other inputs? This service needs to accept a `target` of a NUT type device for one. The user may have HA connected to multiple UPS's. If so it's impossible tell which one to send the command to right now.

But besides that, do any of the commands accept other inputs? If so then I would strongly consider taking the approach I suggested above of mapping the commands to individual services rather then having one generic service. This will get very confusing if you start adding additional fields like `input_1`, `input_2`, etc. to handle the possibility that a command might have additional inputs.

Reply:
Ideally these should be device actions as well since the NUT integration creates a device for each connected UPS.

Reply:
Thanks for the quick response, @mdegat01. Yes, your proposed approach of registering a service for each command is definitely a way that we can go. The main challenge is that the supported commands vary by UPS model, meaning that the list of all supported services is dynamic (although I assume that it is always a subset of those described [here](https://networkupstools.org/docs/user-manual.chunked/apcs03.html)).

What I had in mind was exposing later on a sensor with the list of all the supported commands by each unit, so that it was clear which were the valid values for the command name.

Regarding your proposal, I see a few options, but please let me know if some of them don't make sense as I'm still familiarizing myself with HA:

1. Have a list of common commands and register those as all potential services.
   1. We could also keep the generic `run_command` service as a fallback option for those commands not on the list.
3. Dynamically register services based on the available commands for the units.
   1. That limits our ability to create the `services.yaml` file and I don't know how we would handle having multiple devices with commands that are available to only one of them.
5. Keeping just the existing `run_command` service.

Some commands do accept a single parameter, according to [the documentation](https://networkupstools.org/docs/developer-guide.chunked/ar01s09.html#:~:text=%3Ccmdparam%3E%20is%20an%20additional%20and%20optional%20parameter%20for%20the%20command.), although it is unclear to me which of them are. If we go with the option of the generic service, then we can have a generic optional parameter `cmdparam` or similar. I was planning on tackling this a future enhancement.

What do you think? Which of the three options should we go with?

I will fix the missing target, good catch. I need to test it using a system with multiple UPS devices to ensure that it looks good.

I will also take a look at the device actions. It was not very clear to me from the docs how to use them in this integration, but I will take another pass at it.

Reply:
Is there a way to find out what commands are supported by a particular model by making an API call? If so then that's a strong argument for device actions IMO.

I believe you'll have to register services for all the possible commands to do this and then validate whether the target supports the command based on device metadata from that API call. But what you can do is register different device actions when creating the device. So only the commands that make sense for that model show up in the device action UI making mistakes impossible there.

As for whether to have a fallback generic `run_command` service, I guess it depends how many commands there are. If there's like 20 or so then probably just plan to make the common ones into services now and more over time on user request. If there's hundreds and the list grows regularly then I would agree with a fallback generic `run_command` service as it's unlikely we'll keep up with that. Should also let the code owner weigh in here and they'll be supporting this.

Reply:
> Is there a way to find out what commands are supported by a particular model by making an API call? 

Yes. Device actions would be the best way to go then per your recommendation.

> f there's like 20 [commands] or so then probably just plan to make the common ones into services now and more over time on user request.

That's the case.

> Should also let the code owner weigh in here

@bdraco, @ollo69 - any thoughts?

Reply:
It looks like non of these commands take arguments?

If that is the case they might be a bit more user friendly if you implement them as button entities

Reply:
I guess some of them do take params

```
INSTCMD <upsname> <cmdname> [<cmdparam>]
INSTCMD su700 test.panel.start
INSTCMD su700 load.off.delay 120
```

Reply:
I think we end up with a whole lot of buttons or services and that's not particularly a great experience.

I think device actions is the way to go here.